crates/ide-assists/src/handlers/add_label_to_loop.rs

```diff
@@ -0,0 +1,168 @@
+use ide_db::syntax_helpers::node_ext::for_each_break_and_continue_expr;
+use syntax::ast::{self, AstNode, HasLoopBody};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: add_label_to_loop
+//
+// Adds a label to a loop.
+//
+// ```
+// fn main() {
+//     loop$0 {
+//         break;
+//         continue;
+//     }
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     'l: loop {
+//         break 'l;
+//         continue 'l;
+//     }
+// }
+// ```
+pub(crate) fn add_label_to_loop(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let loop_expr = ctx.find_node_at_offset::<ast::LoopExpr>()?;
+    if loop_expr.label().is_some() {
+        return None;
+    }
+    let loop_body = loop_expr.loop_body().and_then(|it| it.stmt_list());
+    let mut related_exprs = vec![];
+    related_exprs.push(ast::Expr::LoopExpr(loop_expr.clone()));
+    for_each_break_and_continue_expr(loop_expr.label(), loop_body, &mut |expr| {
+        if let ast::Expr::BreakExpr(_) | ast::Expr::ContinueExpr(_) = expr {
+            related_exprs.push(expr)
+        }
+    });
```

Comment:
This can be moved into the `|builder|` closure for lazy evaluation

Reply:
We also don't need to use a vec in that case then

Reply:
Yes, it's better